internal/handlers/common/update.go

```diff
@@ -622,6 +700,57 @@ func extractValueFromUpdateOperator(op string, update *types.Document) (*types.D
 		return nil, NewWriteErrorMsg(ErrFailedToParse, "Modifiers operate on fields but we found another type instead")
 	}
 
+	if op == "$rename" {
+		iter := doc.Iterator()
+		keys := map[string]struct{}{}
+
+		for {
+			k, v, err := iter.Next()
+			if err != nil {
+				if errors.Is(err, iterator.ErrIteratorDone) {
+					break
+				}
+
+				return nil, err
+			}
+
+			var vStr string
+
+			vStr, ok = v.(string)
+			if !ok {
+				return nil, NewWriteErrorMsg(
+					ErrBadValue,
+					fmt.Sprintf("The 'to' field for $rename must be a string: %s: %v", k, vStr),
+				)
+			}
+
+			if k == vStr {
+				return nil, NewWriteErrorMsg(
+					ErrBadValue,
+					fmt.Sprintf("The source and target field for $rename must differ: %s: %v", k, vStr),
+				)
+			}
+
+			if _, ok = keys[k]; ok {
+				return nil, NewWriteErrorMsg(
+					ErrConflictingUpdateOperators,
+					fmt.Sprintf("Updating the '%s' would create a conflict at '%s'", k, k),
+				)
+			}
+
+			keys[k] = struct{}{}
+
+			if _, ok = keys[vStr]; ok {
+				return nil, NewWriteErrorMsg(
+					ErrConflictingUpdateOperators,
```

Comment:
Is it applicable for `$rename`?

Reply:
I'm not sure what do you mean. This part of code is called only for `$rename` operator and it checks for duplicates between keys and values (same as mongodb does). It's not possible to use {$rename: {{"foo": "bar"}, { "bar": "baz"} }} and etc